src/main/java/org/javamodularity/moduleplugin/tasks/PackageScanner.java

```diff
@@ -0,0 +1,95 @@
+package org.javamodularity.moduleplugin.tasks;
+
+import org.codehaus.groovy.tools.Utilities;
+import org.gradle.api.GradleException;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.Opcodes;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Stream;
+
+public class PackageScanner extends ClassVisitor {
```

Comment:
Additionally, is the `asm` dependency really necessary for this? It does mean being tied to a specific bytecode version through the actual version of `asm` used. At the moment it seems they're updating with new Java releases consistently, but that hasn't always been the case and might become a drag in the future. Have you looked at ByteBuddy as an alternative approach (or depending on what's actually happening here just sticking with core Java)?

Reply:
We need to retrieve the list of all test packages in order to pass it to the 'add-opens' option.
There are two ways to do this:
1. Use a parser to analyze the source files of the test code.
2. Analyze the bytecode of the compiled test classes.

The first approach ties us to a specific source code version, the second to a specific bytecode version.
For example, neither JavaParser nor ASM can currently work with Java 12. And ByteBuddy also uses ASM under the hood.

I didn't find a Kotlin parser, therefore I opted for the bytecode analysis approach.

Another idea is to write a mock parser that can only locate the package declaration in a source file and retrieve its value. This would be a version-independent solution, albeit not a very elegant one.

What do you think?

Reply:
I wasn't aware ASM doesn't work with Java 12 yet. That would stop folks from experimenting with newer versions just because of this dependency, that seems a high cost. 

I'm somewhat leaning towards a custom parser, specially if it's really just to get package names.

Reply:
I was probably wrong. It seems there are no bytecode changes planned for Java 12, so ASM 7 should also work with this version.
Nevertheless, I will write a custom parser to get the package names.

Reply:
Awesome, thanks for putting in all the work!